### PR TITLE
mpi: Fix mask ordering for sparse gather

### LINKED ISSUE #1823: Receivers coordinates shuffled when running with higher number of mpi subprocesses
When running the code below for more than 5 receivers, calling the `solver.forward` method shuffles the receiver's coordinates for rank 3 and the assert fails.

```
import numpy as np

from examples.seismic import (demo_model, Receiver, AcquisitionGeometry)
from examples.seismic.acoustic import AcousticWaveSolver


model0 = demo_model('layers-isotropic', vp_top=2.5, vp_bottom=2.5, shape=(101, 101, 101), spacing=(10., 10., 10.), origin=(0., 0., 0.), nbl=40)

source_locations = np.zeros((1, 3))

rec_coordinates = np.array([[1000., 0., 900.], [1000., 300., 700.], [1000., 500., 500.],[1000., 700., 300.], [1000., 900., 0.], [1000., 0., 850.]])

geometry = AcquisitionGeometry(model0, rec_coordinates, source_locations, t0=0., tn=1000., f0=0.010, src_type='Ricker')

solver = AcousticWaveSolver(model0, geometry, space_order=4)

residual = Receiver(name='residual', grid=model0.grid, time_range=geometry.time_axis, coordinates=geometry.rec_positions)
d_syn = Receiver(name='d_syn', grid=model0.grid, time_range=geometry.time_axis, coordinates=geometry.rec_positions)

_, u0, _ = solver.forward(vp=model0.vp, save=False, rec=d_syn)

assert np.allclose(residual.coordinates.data[:], d_syn.coordinates.data)
```

### PR DESCRIPTION
**Pull request by mloubout**
Fixes #1823

### CODE REVIEW

**Review by FabioLuporini — APPROVED** [review]
LGTM. 

@mloubout could you remind me what the reason for the sort was?

**Review by georgebisbas — APPROVED** [review]

File: devito/types/sparse.py [hunk 1]
```diff
@@ -151,37 +151,6 @@ def _dist_scatter_mask(self, dmap=None):
...
-
-        return tuple(ret)
-
-    def _dist_subfunc_gather_mask(self, dmap=None):
-        """
-        This method is analogous to :meth:`_dist_subfunc_scatter_mask`, although
-        the mask is now suitable to index into self's SubFunctions, rather
-        than into ``self.data``.
-        """
-        return self._dist_gather_mask(dmap=dmap)[self._sparse_position]
```
Comment (mloubout) [inline devito/types/sparse.py:183]: Not for reviewer: THis is much "safer" to just reuse the scatter mask reciprocally than trying to infer its inverse. SO gather mask is removed

File: devito/types/sparse.py [hunk 2]
```diff
@@ -151,37 +151,6 @@ def _dist_scatter_mask(self, dmap=None):
         ret[self._sparse_position] = mask
         return tuple(ret)
 
-    def _dist_subfunc_scatter_mask(self, dmap=None):
-        """
-        This method is analogous to :meth:`_dist_scatter_mask`, although
-        the mask is now suitable to index into self's SubFunctions, rather
-        than into ``self.data``.
-        """
-        return self._dist_scatter_mask(dmap=dmap)[self._sparse_position]
```
Comment (mloubout) [inline devito/types/sparse.py:160]: Note for reviewer: Slightly cheaper to just compute the mask and access sparse position than recopute it just to access the sparse position.

**Review by mloubout — COMMENTED** [review]
@FabioLuporini Should be GTG now and less likely to break in future (and slightly more efficient)

**Review by georgebisbas — APPROVED** [review]
Good to see code dropped..

**Review by FabioLuporini — APPROVED** [review]
LGTM, thanks

**Pull request merged**